File: frontend/src/stores/obeya.ts
import { create } from 'zustand';
import { devtools, persist } from 'zustand/middleware';

type ObeyaCategory = 
  | 'issue' | 'action' | 'risk' | 'decision' | 'milestone' 
  | 'kpi' | 'escalation' | 'information' | 'lesson_learned'
  | 'metrics' | 'schedule' | 'quality' | 'cost' | 'safety' 
  | 'morale' | 'delivery' | 'strategy';

type ObeyaStatus = 'new' | 'in_progress' | 'blocked' | 'waiting' | 'completed' | 'cancelled';
type ObeyaPriority = 'low' | 'medium' | 'high' | 'critical';
type ObeyaBoard = 'daily' | 'weekly' | 'project' | 'strategic' | 'quality' | 'safety' | 'improvement';

interface ObeyaItem {
  id: string;
  board: ObeyaBoard;
  column?: string;
  position: number;
  title: string;
  description?: string;
  category: ObeyaCategory;
  status: ObeyaStatus;
  priority: ObeyaPriority;
  color?: string;
  related_entity_type?: string;
  related_entity_id?: string;
  assigned_to_id?: string;
  assigned_to_name?: string;
  due_date?: string;
  target_date?: string;
  completed_at?: string;
  blocked_reason?: string;
  resolution?: string;
  decision_outcome?: string;
  decision_rationale?: string;
  kpi_target?: string;
  kpi_actual?: string;
  kpi_unit?: string;
  kpi_trend?: 'improving' | 'stable' | 'declining';
  is_escalated: boolean;
  escalated_to_id?: string;
  escalated_to_name?: string;
  escalated_at?: string;
  escalation_reason?: string;
  days_open?: number;
  days_overdue?: number;
  attachments?: any[];
  links?: any[];
  tags?: string[];
  meeting_date?: string;
  meeting_type?: string;
  notes?: string;
  custom_fields?: Record<string, any>;
  comments_count: number;
  created_at: string;
  updated_at: string;
  created_by_id: string;
  created_by_name: string;
}

interface ObeyaComment {
  id: string;
  item_id: string;
  author_id: string;
  author_name: string;
  content: string;
  parent_id?: string;
  is_status_change: boolean;
  old_status?: string;
  new_status?: string;
  is_pinned: boolean;
  is_edited: boolean;
  edited_at?: string;
  mentions?: string[];
  attachments?: any[];
  created_at: string;
}

interface ObeyaStats {
  total_items: number;
  by_status: Record<ObeyaStatus, number>;
  by_priority: Record<ObeyaPriority, number>;
  by_category: Record<ObeyaCategory, number>;
  overdue_count: number;
  escalated_count: number;
  completed_this_week: number;
  avg_days_to_complete: number;
}

interface SQDCPMetrics {
  safety: {
    incidents: number;
    days_since_last_incident: number;
    near_misses: number;
    training_completion: number;
    status: 'green' | 'yellow' | 'red';
  };
  quality: {
    first_pass_yield: number;
    defect_rate: number;
    customer_complaints: number;
    ncr_open: number;
    status: 'green' | 'yellow' | 'red';
  };
  delivery: {
    on_time_delivery: number;
    lead_time_days: number;
    schedule_adherence: number;
    backlog_items: number;
    status: 'green' | 'yellow' | 'red';
  };
  cost: {
    variance_percent: number;
    cost_savings: number;
    waste_reduction: number;
    budget_utilization: number;
    status: 'green' | 'yellow' | 'red';
  };
  people: {
    morale_score: number;
    training_hours: number;
    attendance_rate: number;
    active_improvements: number;
    status: 'green' | 'yellow' | 'red';
  };
}

interface ObeyaState {
  items: ObeyaItem[];
  stats: ObeyaStats;
  sqdcpMetrics: SQDCPMetrics | null;
  cognitiveInsights: any | null;
  selectedBoard: ObeyaBoard;
  isLoading: boolean;
  error: string | null;
  lastFetchedAt: number | null;
  socket: WebSocket | null;
  isConnected: boolean;

  fetchItems: (board?: ObeyaBoard) => Promise<void>;
  fetchItemById: (id: string) => Promise<ObeyaItem | null>;
  fetchSQDCPMetrics: () => Promise<void>;
  fetchCognitiveInsights: () => Promise<void>;
  createItem: (item: Partial<ObeyaItem>) => Promise<ObeyaItem>;
  updateItem: (id: string, updates: Partial<ObeyaItem>) => Promise<ObeyaItem>;
  deleteItem: (id: string) => Promise<void>;
  moveItem: (id: string, column: string, position: number) => Promise<void>;
  addComment: (itemId: string, comment: Partial<ObeyaComment>) => Promise<ObeyaComment>;
  fetchComments: (itemId: string) => Promise<ObeyaComment[]>;
  escalateItem: (id: string, reason: string, escalatedToId: string) => Promise<void>;
  resolveItem: (id: string, resolution: string) => Promise<void>;
  setSelectedBoard: (board: ObeyaBoard) => void;
  clearError: () => void;
  connect: () => void;
  disconnect: () => void;
}

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api/v1';

const initialStats: ObeyaStats = {
  total_items: 0,
  by_status: {
    new: 0,
    in_progress: 0,
    blocked: 0,
    waiting: 0,
    completed: 0,
    cancelled: 0,
  },
  by_priority: {
    low: 0,
    medium: 0,
    high: 0,
    critical: 0,
  },
  by_category: {
    issue: 0,
    action: 0,
    risk: 0,
    decision: 0,
    milestone: 0,
    kpi: 0,
    escalation: 0,
    information: 0,
    lesson_learned: 0,
    metrics: 0,
    schedule: 0,
    quality: 0,
    cost: 0,
    safety: 0,
    morale: 0,
    delivery: 0,
    strategy: 0,
  },
  overdue_count: 0,
  escalated_count: 0,
  completed_this_week: 0,
  avg_days_to_complete: 0,
};

export const useObeyaStore = create<ObeyaState>()(
  devtools(
    persist(
      (set, get) => ({
        items: [],
        stats: initialStats,
        sqdcpMetrics: null,
        selectedBoard: 'daily',
        isLoading: false,
        error: null,
        lastFetchedAt: null,
        socket: null,
        isConnected: false,

        connect: () => {
          const { socket, isConnected } = get();
          if (socket || isConnected) return;

          const token = localStorage.getItem('access_token');
          if (!token) return;

          const wsUrl = API_BASE_URL.replace('http', 'ws').replace('/api/v1', '/ws');
          const newSocket = new WebSocket(`${wsUrl}/${token}`);

          newSocket.onopen = () => {
            set({ isConnected: true });
            console.log('Obeya WebSocket connected');
          };

          newSocket.onmessage = (event) => {
            try {
              const data = JSON.parse(event.data);
              if (data.type === 'metric_update') {
                // Refresh metrics when they change
                get().fetchSQDCPMetrics();
                get().fetchCognitiveInsights();
              }
              if (data.type === 'silo_alert') {
                get().fetchCognitiveInsights();
              }
            } catch (e) {
              console.error('Error parsing Obeya WebSocket message:', e);
            }
          };

          newSocket.onclose = () => {
            set({ isConnected: false, socket: null });
            console.log('Obeya WebSocket disconnected');
            // Try to reconnect after 5 seconds
            setTimeout(() => get().connect(), 5000);
          };

          set({ socket: newSocket });
        },

        disconnect: () => {
          const { socket } = get();
          if (socket) {
            socket.close();
          }
          set({ socket: null, isConnected: false });
        },

        fetchItems: async (board?: ObeyaBoard) => {
          const { lastFetchedAt, selectedBoard } = get();
          const now = Date.now();
          const targetBoard = board || selectedBoard;

          // Cache for 30 seconds
          if (lastFetchedAt && now - lastFetchedAt < 30000) {
            return;
          }

          set({ isLoading: true, error: null });
          try {
            const response = await fetch(`${API_BASE_URL}/obeya/items?board=${targetBoard}`, {
              headers: {
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
              },
            });

            if (!response.ok) {
              throw new Error(`Failed to fetch Obeya items: ${response.statusText}`);
            }

            const data = await response.json();
            const items: ObeyaItem[] = data.items || [];

            // Calculate stats
            const stats: ObeyaStats = {
              total_items: items.length,
              by_status: items.reduce((acc, item) => {
                acc[item.status] = (acc[item.status] || 0) + 1;
                return acc;
              }, {} as Record<ObeyaStatus, number>),
              by_priority: items.reduce((acc, item) => {
                acc[item.priority] = (acc[item.priority] || 0) + 1;
                return acc;
              }, {} as Record<ObeyaPriority, number>),
              by_category: items.reduce((acc, item) => {
                acc[item.category] = (acc[item.category] || 0) + 1;
                return acc;
              }, {} as Record<ObeyaCategory, number>),
              overdue_count: items.filter(i => i.days_overdue && i.days_overdue > 0).length,
              escalated_count: items.filter(i => i.is_escalated).length,
              completed_this_week: items.filter(i => {
                if (!i.completed_at) return false;
                const completedDate = new Date(i.completed_at);
                const weekAgo = new Date();
                weekAgo.setDate(weekAgo.getDate() - 7);
                return completedDate >= weekAgo;
              }).length,
              avg_days_to_complete: 0, // Calculated on backend
            };

            set({
              items,
              stats,
              selectedBoard: targetBoard,
              lastFetchedAt: now,
              isLoading: false,
            });
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to fetch Obeya items',
              isLoading: false,
            });
          }
        },

        fetchItemById: async (id: string) => {
          set({ isLoading: true, error: null });
          try {
            const response = await fetch(`${API_BASE_URL}/obeya/items/${id}`, {
              headers: {
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
              },
            });

            if (!response.ok) {
              throw new Error(`Failed to fetch Obeya item: ${response.statusText}`);
            }

            const item: ObeyaItem = await response.json();

            set(state => ({
              items: state.items.map(i => i.id === id ? item : i),
              isLoading: false,
            }));

            return item;
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to fetch Obeya item',
              isLoading: false,
            });
            return null;
          }
        },

        fetchSQDCPMetrics: async () => {
          set({ isLoading: true, error: null });
          try {
            const response = await fetch(`${API_BASE_URL}/obeya/sqdcp-metrics`, {
              headers: {
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
              },
            });

            if (!response.ok) {
              throw new Error(`Failed to fetch SQDCP metrics: ${response.statusText}`);
            }

            const metrics: SQDCPMetrics = await response.json();

            set({ sqdcpMetrics: metrics, isLoading: false });
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to fetch SQDCP metrics',
              isLoading: false,
            });
          }
        },

        fetchCognitiveInsights: async () => {
          set({ isLoading: true, error: null });
          try {
            const response = await fetch(`${API_BASE_URL}/cognitive-obeya/dashboard`, {
              headers: {
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
              },
            });

            if (!response.ok) {
              throw new Error(`Failed to fetch Cognitive insights: ${response.statusText}`);
            }

            const data = await response.json();
            set({ cognitiveInsights: data.data, isLoading: false });
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to fetch Cognitive insights',
              isLoading: false,
            });
          }
        },

        createItem: async (itemData: Partial<ObeyaItem>) => {
          set({ isLoading: true, error: null });
          try {
            const response = await fetch(`${API_BASE_URL}/obeya/items`, {
              method: 'POST',
              headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
              },
              body: JSON.stringify(itemData),
            });

            if (!response.ok) {
              throw new Error(`Failed to create Obeya item: ${response.statusText}`);
            }

            const item: ObeyaItem = await response.json();

            set(state => ({
              items: [item, ...state.items],
              stats: {
                ...state.stats,
                total_items: state.stats.total_items + 1,
              },
              isLoading: false,
            }));

            return item;
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to create Obeya item',
              isLoading: false,
            });
            throw error;
          }
        },

        updateItem: async (id: string, updates: Partial<ObeyaItem>) => {
          set({ isLoading: true, error: null });
          try {
            const response = await fetch(`${API_BASE_URL}/obeya/items/${id}`, {
              method: 'PATCH',
              headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
              },
              body: JSON.stringify(updates),
            });

            if (!response.ok) {
              throw new Error(`Failed to update Obeya item: ${response.statusText}`);
            }

            const item: ObeyaItem = await response.json();

            set(state => ({
              items: state.items.map(i => i.id === id ? item : i),
              isLoading: false,
            }));

            return item;
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to update Obeya item',
              isLoading: false,
            });
            throw error;
          }
        },

        deleteItem: async (id: string) => {
          set({ isLoading: true, error: null });
          try {
            const response = await fetch(`${API_BASE_URL}/obeya/items/${id}`, {
              method: 'DELETE',
              headers: {
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
              },
            });

            if (!response.ok) {
              throw new Error(`Failed to delete Obeya item: ${response.statusText}`);
            }

            set(state => ({
              items: state.items.filter(i => i.id !== id),
              stats: {
                ...state.stats,
                total_items: Math.max(0, state.stats.total_items - 1),
              },
              isLoading: false,
            }));
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to delete Obeya item',
              isLoading: false,
            });
            throw error;
          }
        },

        moveItem: async (id: string, column: string, position: number) => {
          set({ isLoading: true, error: null });
          try {
            const response = await fetch(`${API_BASE_URL}/obeya/items/${id}/move`, {
              method: 'PATCH',
              headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
              },
              body: JSON.stringify({ column, position }),
            });

            if (!response.ok) {
              throw new Error(`Failed to move Obeya item: ${response.statusText}`);
            }

            const item: ObeyaItem = await response.json();

            set(state => ({
              items: state.items.map(i => i.id === id ? item : i),
              isLoading: false,
            }));
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to move Obeya item',
              isLoading: false,
            });
            throw error;
          }
        },

        addComment: async (itemId: string, commentData: Partial<ObeyaComment>) => {
          set({ isLoading: true, error: null });
          try {
            const response = await fetch(`${API_BASE_URL}/obeya/items/${itemId}/comments`, {
              method: 'POST',
              headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
              },
              body: JSON.stringify(commentData),
            });

            if (!response.ok) {
              throw new Error(`Failed to add comment: ${response.statusText}`);
            }

            const comment: ObeyaComment = await response.json();

            // Update item comment count
            set(state => ({
              items: state.items.map(i => 
                i.id === itemId 
                  ? { ...i, comments_count: i.comments_count + 1 }
                  : i
              ),
              isLoading: false,
            }));

            return comment;
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to add comment',
              isLoading: false,
            });
            throw error;
          }
        },

        fetchComments: async (itemId: string) => {
          set({ isLoading: true, error: null });
          try {
            const response = await fetch(`${API_BASE_URL}/obeya/items/${itemId}/comments`, {
              headers: {
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
              },
            });

            if (!response.ok) {
              throw new Error(`Failed to fetch comments: ${response.statusText}`);
            }

            const comments: ObeyaComment[] = await response.json();
            set({ isLoading: false });
            return comments;
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to fetch comments',
              isLoading: false,
            });
            return [];
          }
        },

        escalateItem: async (id: string, reason: string, escalatedToId: string) => {
          set({ isLoading: true, error: null });
          try {
            const response = await fetch(`${API_BASE_URL}/obeya/items/${id}/escalate`, {
              method: 'POST',
              headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
              },
              body: JSON.stringify({ reason, escalated_to_id: escalatedToId }),
            });

            if (!response.ok) {
              throw new Error(`Failed to escalate item: ${response.statusText}`);
            }

            const item: ObeyaItem = await response.json();

            set(state => ({
              items: state.items.map(i => i.id === id ? item : i),
              stats: {
                ...state.stats,
                escalated_count: state.stats.escalated_count + 1,
              },
              isLoading: false,
            }));
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to escalate item',
              isLoading: false,
            });
            throw error;
          }
        },

        resolveItem: async (id: string, resolution: string) => {
          set({ isLoading: true, error: null });
          try {
            const response = await fetch(`${API_BASE_URL}/obeya/items/${id}/resolve`, {
              method: 'POST',
              headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
              },
              body: JSON.stringify({ resolution }),
            });

            if (!response.ok) {
              throw new Error(`Failed to resolve item: ${response.statusText}`);
            }

            const item: ObeyaItem = await response.json();

            set(state => ({
              items: state.items.map(i => i.id === id ? item : i),
              isLoading: false,
            }));
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to resolve item',
              isLoading: false,
            });
            throw error;
          }
        },

        setSelectedBoard: (board: ObeyaBoard) => {
          set({ selectedBoard: board });
          get().fetchItems(board);
        },

        clearError: () => set({ error: null }),
      }),
      {
        name: 'obeya-storage',
        partialize: (state) => ({
          items: state.items,
          stats: state.stats,
          sqdcpMetrics: state.sqdcpMetrics,
          cognitiveInsights: state.cognitiveInsights,
          selectedBoard: state.selectedBoard,
          lastFetchedAt: state.lastFetchedAt,
        }),
      }
    )
  )
);

File: sensei-rs/crates/sensei-frontend/src/stores/obeya.rs
//! Obeya board store — items, comments, SQDCP metrics, cognitive insights,
//! and real-time WebSocket updates.
//!
//! Port of [`frontend/src/stores/obeya.ts`](frontend/src/stores/obeya.ts).

use leptos::prelude::*;
use std::collections::HashMap;
use crate::api::client::{ApiClient, ApiError};

// ---------------------------------------------------------------------------
// Domain types
// ---------------------------------------------------------------------------

pub type ObeyaBoard = String;
pub type ObeyaCategory = String;
pub type ObeyaPriority = String;
pub type ObeyaStatus = String;

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ObeyaItem {
    pub id: String,
    pub board_id: String,
    pub title: String,
    pub description: String,
    pub category: ObeyaCategory,
    pub priority: ObeyaPriority,
    pub status: ObeyaStatus,
    pub column: String,
    pub position: i32,
    pub assigned_to: Option<String>,
    pub due_date: Option<String>,
    pub escalated: bool,
    pub escalated_to: Option<String>,
    pub escalation_reason: Option<String>,
    pub resolution: Option<String>,
    pub tags: Vec<String>,
    pub created_by: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ObeyaComment {
    pub id: String,
    pub item_id: String,
    pub author: String,
    pub content: String,
    pub created_at: String,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ObeyaStats {
    pub total_items: i32,
    pub by_status: HashMap<String, i32>,
    pub by_priority: HashMap<String, i32>,
    pub by_category: HashMap<String, i32>,
    pub completed_this_week: i32,
    pub overdue: i32,
}

impl Default for ObeyaStats {
    fn default() -> Self {
        Self {
            total_items: 0,
            by_status: HashMap::new(),
            by_priority: HashMap::new(),
            by_category: HashMap::new(),
            completed_this_week: 0,
            overdue: 0,
        }
    }
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct SqdqpMetrics {
    pub quality: f64,
    pub delivery: f64,
    pub safety: f64,
    pub morale: f64,
    pub productivity: f64,
    pub cost: f64,
    pub overall: f64,
    pub trend: String,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct CognitiveInsight {
    pub id: String,
    pub insight_type: String,
    pub title: String,
    pub description: String,
    pub severity: String,
    pub recommendation: Option<String>,
    pub created_at: String,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ObeyaComments {
    pub item_id: String,
    pub comments: Vec<ObeyaComment>,
}

// ---------------------------------------------------------------------------
// ObeyaStore
// ---------------------------------------------------------------------------

#[derive(Debug, Clone)]
pub struct ObeyaStore {
    // Data signals
    pub items: RwSignal<Vec<ObeyaItem>>,
    pub current_item: RwSignal<Option<ObeyaItem>>,
    pub comments: RwSignal<HashMap<String, Vec<ObeyaComment>>>,
    pub stats: RwSignal<Option<ObeyaStats>>,
    pub sqdcp_metrics: RwSignal<Option<SqdqpMetrics>>,
    pub cognitive_insights: RwSignal<Vec<CognitiveInsight>>,
    pub selected_board: RwSignal<Option<ObeyaBoard>>,
    pub is_connected: RwSignal<bool>,

    // Loading & error
    pub loading: RwSignal<bool>,
    pub error: RwSignal<Option<String>>,
}

impl ObeyaStore {
    pub fn new() -> Self {
        Self {
            items: RwSignal::new(Vec::new()),
            current_item: RwSignal::new(None),
            comments: RwSignal::new(HashMap::new()),
            stats: RwSignal::new(None),
            sqdcp_metrics: RwSignal::new(None),
            cognitive_insights: RwSignal::new(Vec::new()),
            selected_board: RwSignal::new(None),
            is_connected: RwSignal::new(false),
            loading: RwSignal::new(false),
            error: RwSignal::new(None),
        }
    }

    pub fn set_selected_board(&self, board: ObeyaBoard) {
        self.selected_board.set(Some(board));
    }

    pub fn clear_error(&self) {
        self.error.set(None);
    }

    // Compute stats from items
    fn compute_stats(&self) -> ObeyaStats {
        let items = self.items.get();
        let total_items = items.len() as i32;

        let mut by_status: HashMap<String, i32> = HashMap::new();
        let mut by_priority: HashMap<String, i32> = HashMap::new();
        let mut by_category: HashMap<String, i32> = HashMap::new();
        let mut completed_this_week = 0;
        let mut overdue = 0;

        for item in items.iter() {
            *by_status.entry(item.status.clone()).or_insert(0) += 1;
            *by_priority.entry(item.priority.clone()).or_insert(0) += 1;
            *by_category.entry(item.category.clone()).or_insert(0) += 1;

            if item.status == "completed" {
                completed_this_week += 1;
            }
            if let Some(ref due) = item.due_date {
                // Simple check — if due date is in the past and not completed
                if due.as_str() < "2099-01-01" && item.status != "completed" {
                    overdue += 1;
                }
            }
        }

        ObeyaStats {
            total_items,
            by_status,
            by_priority,
            by_category,
            completed_this_week,
            overdue,
        }
    }

    // -----------------------------------------------------------------------
    // Items
    // -----------------------------------------------------------------------

    pub async fn fetch_items(&self, client: &ApiClient, board: Option<&str>) {
        self.loading.set(true);
        self.error.set(None);
        let path = match board {
            Some(b) => format!("/obeya/items?board={b}"),
            None => "/obeya/items".to_string(),
        };
        match client.get::<Vec<ObeyaItem>>(&path).await {
            Ok(items) => {
                self.items.set(items);
                let stats = self.compute_stats();
                self.stats.set(Some(stats));
            }
            Err(e) => self.error.set(Some(e.to_string())),
        }
        self.loading.set(false);
    }

    pub async fn fetch_item_by_id(&self, client: &ApiClient, id: &str) {
        self.loading.set(true);
        match client.get::<ObeyaItem>(&format!("/obeya/items/{id}")).await {
            Ok(item) => self.current_item.set(Some(item)),
            Err(e) => self.error.set(Some(e.to_string())),
        }
        self.loading.set(false);
    }

    pub async fn create_item(&self, client: &ApiClient, item_data: serde_json::Value) -> Result<ObeyaItem, ()> {
        self.loading.set(true);
        match client.post::<ObeyaItem, serde_json::Value>("/obeya/items", &item_data).await {
            Ok(new_item) => {
                self.items.update(|items| items.push(new_item.clone()));
                let stats = self.compute_stats();
                self.stats.set(Some(stats));
                self.loading.set(false);
                Ok(new_item)
            }
            Err(e) => {
                self.error.set(Some(e.to_string()));
                self.loading.set(false);
                Err(())
            }
        }
    }

    pub async fn update_item(&self, client: &ApiClient, id: &str, updates: serde_json::Value) -> Result<ObeyaItem, ()> {
        self.loading.set(true);
        match client.put::<ObeyaItem, serde_json::Value>(&format!("/obeya/items/{id}"), &updates).await {
            Ok(updated) => {
                self.items.update(|items| {
                    if let Some(pos) = items.iter().position(|i| i.id == id) {
                        items[pos] = updated.clone();
                    }
                });
                self.current_item.set(Some(updated.clone()));
                let stats = self.compute_stats();
                self.stats.set(Some(stats));
                self.loading.set(false);
                Ok(updated)
            }
            Err(e) => {
                self.error.set(Some(e.to_string()));
                self.loading.set(false);
                Err(())
            }
        }
    }

    pub async fn delete_item(&self, client: &ApiClient, id: &str) -> Result<(), ()> {
        self.loading.set(true);
        match client.delete::<serde_json::Value>(&format!("/obeya/items/{id}")).await {
            Ok(_) => {
                self.items.update(|items| items.retain(|i| i.id != id));
                let stats = self.compute_stats();
                self.stats.set(Some(stats));
                self.loading.set(false);
                Ok(())
            }
            Err(e) => {
                self.error.set(Some(e.to_string()));
                self.loading.set(false);
                Err(())
            }
        }
    }

    pub async fn move_item(&self, client: &ApiClient, id: &str, column: &str, position: i32) -> Result<ObeyaItem, ()> {
        let updates = serde_json::json!({ "column": column, "position": position });
        self.update_item(client, id, updates).await
    }

    // -----------------------------------------------------------------------
    // Comments
    // -----------------------------------------------------------------------

    pub async fn add_comment(&self, client: &ApiClient, item_id: &str, comment_data: serde_json::Value) -> Result<ObeyaComment, ()> {
        match client.post::<ObeyaComment, serde_json::Value>(
            &format!("/obeya/items/{item_id}/comments"), &comment_data,
        ).await {
            Ok(comment) => {
                self.comments.update(|c| {
                    c.entry(item_id.to_string()).or_default().push(comment.clone());
                });
                Ok(comment)
            }
            Err(e) => {
                self.error.set(Some(e.to_string()));
                Err(())
            }
        }
    }

    pub async fn fetch_comments(&self, client: &ApiClient, item_id: &str) {
        match client.get::<Vec<ObeyaComment>>(&format!("/obeya/items/{item_id}/comments")).await {
            Ok(comments) => {
                self.comments.update(|c| {
                    c.insert(item_id.to_string(), comments);
                });
            }
            Err(e) => self.error.set(Some(e.to_string())),
        }
    }

    // -----------------------------------------------------------------------
    // Escalation & Resolution
    // -----------------------------------------------------------------------

    pub async fn escalate_item(&self, client: &ApiClient, id: &str, reason: &str, escalated_to_id: &str) -> Result<ObeyaItem, ()> {
        let updates = serde_json::json!({
            "escalated": true,
            "escalation_reason": reason,
            "escalated_to": escalated_to_id,
            "status": "escalated",
        });
        self.update_item(client, id, updates).await
    }

    pub async fn resolve_item(&self, client: &ApiClient, id: &str, resolution: &str) -> Result<ObeyaItem, ()> {
        let updates = serde_json::json!({
            "resolution": resolution,
            "status": "resolved",
            "escalated": false,
        });
        self.update_item(client, id, updates).await
    }

    // -----------------------------------------------------------------------
    // SQDCP Metrics & Cognitive Insights
    // -----------------------------------------------------------------------

    pub async fn fetch_sqdcp_metrics(&self, client: &ApiClient) {
        self.loading.set(true);
        match client.get::<SqdqpMetrics>(&format!("/obeya/sqdcp")).await {
            Ok(metrics) => self.sqdcp_metrics.set(Some(metrics)),
            Err(e) => self.error.set(Some(e.to_string())),
        }
        self.loading.set(false);
    }

    pub async fn fetch_cognitive_insights(&self, client: &ApiClient) {
        self.loading.set(true);
        match client.get::<Vec<CognitiveInsight>>(&format!("/obeya/cognitive-insights")).await {
            Ok(insights) => self.cognitive_insights.set(insights),
            Err(e) => self.error.set(Some(e.to_string())),
        }
        self.loading.set(false);
    }
}

impl Default for ObeyaStore {
    fn default() -> Self {
        Self::new()
    }
}
